modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/registry/AbstractRegistry.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * (C) Copyright 2020 Nuxeo (http://nuxeo.com/) and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Contributors:
+ *     Anahide Tchertchian
+ */
+package org.nuxeo.common.xmap.registry;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+
+import org.nuxeo.common.xmap.Context;
+import org.nuxeo.common.xmap.XAnnotatedObject;
+import org.w3c.dom.Element;
+
+/**
+ * @since TODO
+ */
+public abstract class AbstractRegistry implements Registry {
+
+    protected boolean initialized = false;
+
+    protected Set<String> flags = new HashSet<>();
+
+    protected List<RegistryContribution> registrations = new ArrayList<>();
+
+    public AbstractRegistry() {
+    }
+
+    @Override
+    public boolean isNull() {
+        return false;
+    }
+
+    public boolean isInitialized() {
+        return initialized;
+    }
+
+    public void setInitialized(boolean initialized) {
+        this.initialized = initialized;
+    }
+
+    protected void checkInitialized() {
+        if (isInitialized()) {
+            return;
+        }
+        initialize();
+    }
+
+    protected void initialize() {
+        registrations.forEach(rc -> register(rc.getContext(), rc.getObject(), rc.getElement()));
+        setInitialized(true);
+    }
+
+    @Override
+    public void flag(String id) {
+        flags.add(id);
+    }
+
+    @Override
+    public boolean isFlagged(String id) {
+        return flags.contains(id);
+    }
+
+    @Override
+    public void register(Context ctx, XAnnotatedObject xObject, Element element, String flag) {
+        flags.add(flag);
+        registrations.add(new RegistryContribution(ctx, xObject, element, flag));
```

Comment:
The idea here is to call the other register method (that will process DOM and apply merge logics and such) only when the registry getters are called (at initialization).
It should not hurt to process these right away because they should be retrieved at some point anyway, but in practice it is useful for hot reload logics (so that computations are not done on the fly, some of them being potentially not needed depending on the deploy/undeploy subsequent calls)

Reply:
I kept the logic as is in #4537, please tell if you still think there should be changes in this case